File: dashboard/src/types/address.ts
/**
 * TypeScript type definitions for Address entities
 * Feature: User Dashboard for Wallet Management
 * Generated: 2025-10-17
 */

/**
 * Blockchain category classification
 */
export enum Category {
  /** Original Bitcoin and Layer-1 blockchains */
  BASE = 'base',

  /** Ethereum Layer-2 scaling solutions */
  LAYER2 = 'layer2',

  /** Region-specific or local blockchains */
  REGIONAL = 'regional',

  /** Cosmos ecosystem chains */
  COSMOS = 'cosmos',

  /** Alternative EVM-compatible chains */
  ALT_EVM = 'alt_evm',

  /** Specialized or niche blockchains */
  SPECIALIZED = 'specialized',
}

/**
 * Cryptographic key type for address derivation
 */
export enum KeyType {
  /** ECDSA secp256k1 (Bitcoin, Ethereum) */
  SECP256K1 = 'secp256k1',

  /** EdDSA Ed25519 (Solana, Cardano, Polkadot) */
  ED25519 = 'ed25519',

  /** Schnorr sr25519 (Substrate-based chains) */
  SR25519 = 'sr25519',

  /** Schnorr signature scheme (Bitcoin Taproot) */
  SCHNORR = 'schnorr',
}

/**
 * Represents a derived cryptocurrency address for a specific blockchain
 * Each wallet contains exactly 54 addresses (one per supported blockchain)
 */
export interface Address {
  /** Parent wallet identifier (links to Wallet.id) */
  wallet_id: string;

  /** Display order (1-54, matches SLIP-44 registration order) */
  rank: number;

  /** Blockchain symbol (e.g., "BTC", "ETH", "SOL") */
  symbol: string;

  /** Human-readable blockchain name (e.g., "Bitcoin", "Ethereum") */
  name: string;

  /** SLIP-44 coin type (e.g., 0 for BTC, 60 for ETH) */
  coin_type: number;

  /** BIP44 derivation path (e.g., "m/44'/0'/0'/0/0") */
  derivation_path: string;

  /** Derived public address (blockchain-specific format) */
  address: string;

  /** Blockchain category for filtering */
  category: Category;

  /** Cryptographic key type used for derivation */
  key_type: KeyType;

  /** Optional testnet indicator (true for testnet addresses) */
  is_testnet?: boolean;
}

/**
 * Response from load_addresses Tauri command
 * Contains all 54 derived addresses for a wallet
 */
export interface AddressListResponse {
  /** Wallet identifier */
  wallet_id: string;

  /** Array of all 54 blockchain addresses */
  addresses: Address[];

  /** Total address count (always 54 for v0.3.0) */
  total_count: number;
}

/**
 * Parameters for loading wallet addresses
 */
export interface LoadAddressesParams {
  /** Wallet identifier */
  wallet_id: string;

  /** Wallet encryption password */
  password: string;

  /** USB mount point */
  usb_path: string;
}

/**
 * Filter criteria for address list display
 */
export interface AddressFilter {
  /** Filter by blockchain category (optional) */
  category?: Category;

  /** Filter by key type (optional) */
  key_type?: KeyType;

  /** Search query for symbol or name (optional) */
  search?: string;

  /** Show only testnet addresses (optional) */
  testnet_only?: boolean;
}

/**
 * Export format options
 */
export enum ExportFormat {
  /** JSON format with full metadata */
  JSON = 'json',

  /** CSV format (symbol, address, derivation_path) */
  CSV = 'csv',
}

/**
 * Parameters for exporting addresses
 */
export interface ExportAddressesParams {
  /** Wallet identifier */
  wallet_id: string;

  /** Export file format */
  format: ExportFormat;

  /** Optional filter to export subset of addresses */
  filter?: AddressFilter;
}

/**
 * Response from export_addresses Tauri command
 */
export interface ExportResponse {
  /** Export file path */
  file_path: string;

  /** Number of addresses exported */
  exported_count: number;

  /** Export format used */
  format: ExportFormat;

  /** Export timestamp (ISO 8601) */
  exported_at: string;
}

File: dashboard/src-tauri/src/models/address.rs
/**
 * Address domain model
 * Feature: User Dashboard for Wallet Management
 * Task: T051 - Create Address model
 * Generated: 2025-10-17
 */

use serde::{Deserialize, Serialize};

/// Blockchain category classification
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum Category {
    Base,
    Layer2,
    Regional,
    Cosmos,
    #[serde(rename = "alt_evm")]
    AltEvm,
    Specialized,
}

/// Cryptographic key type for address derivation
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum KeyType {
    Secp256k1,
    Ed25519,
    Sr25519,
    Schnorr,
}

/// Derived cryptocurrency address for a specific blockchain
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Address {
    /// Parent wallet identifier
    pub wallet_id: String,

    /// Display order (1-54)
    pub rank: u32,

    /// Blockchain symbol (e.g., "BTC", "ETH")
    pub symbol: String,

    /// Human-readable blockchain name
    pub name: String,

    /// SLIP-44 coin type
    pub coin_type: u32,

    /// BIP44 derivation path
    pub derivation_path: String,

    /// Derived public address
    pub address: String,

    /// Blockchain category
    pub category: Category,

    /// Cryptographic key type
    pub key_type: KeyType,

    /// Optional testnet indicator
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_testnet: Option<bool>,
}

/// Response from load_addresses command
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AddressListResponse {
    /// Wallet identifier
    pub wallet_id: String,

    /// Array of all 54 blockchain addresses
    pub addresses: Vec<Address>,

    /// Total address count (always 54 for v0.3.0)
    pub total_count: u32,
}

impl Address {
    /// Create new Address instance
    pub fn new(
        wallet_id: String,
        rank: u32,
        symbol: String,
        name: String,
        coin_type: u32,
        derivation_path: String,
        address: String,
        category: Category,
        key_type: KeyType,
    ) -> Self {
        Self {
            wallet_id,
            rank,
            symbol,
            name,
            coin_type,
            derivation_path,
            address,
            category,
            key_type,
            is_testnet: None,
        }
    }

    /// Set testnet flag
    pub fn with_testnet(mut self, is_testnet: bool) -> Self {
        self.is_testnet = Some(is_testnet);
        self
    }
}

impl AddressListResponse {
    /// Create new address list response
    pub fn new(wallet_id: String, addresses: Vec<Address>) -> Self {
        let total_count = addresses.len() as u32;
        Self {
            wallet_id,
            addresses,
            total_count,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_address_creation() {
        let address = Address::new(
            "wallet-1".to_string(),
            1,
            "BTC".to_string(),
            "Bitcoin".to_string(),
            0,
            "m/44'/0'/0'/0/0".to_string(),
            "1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa".to_string(),
            Category::Base,
            KeyType::Secp256k1,
        );

        assert_eq!(address.wallet_id, "wallet-1");
        assert_eq!(address.rank, 1);
        assert_eq!(address.symbol, "BTC");
        assert_eq!(address.category, Category::Base);
        assert_eq!(address.key_type, KeyType::Secp256k1);
    }

    #[test]
    fn test_address_with_testnet() {
        let address = Address::new(
            "wallet-1".to_string(),
            1,
            "BTC".to_string(),
            "Bitcoin".to_string(),
            0,
            "m/44'/0'/0'/0/0".to_string(),
            "address".to_string(),
            Category::Base,
            KeyType::Secp256k1,
        )
        .with_testnet(true);

        assert_eq!(address.is_testnet, Some(true));
    }

    #[test]
    fn test_category_serialization() {
        let category = Category::Base;
        let json = serde_json::to_string(&category).unwrap();
        assert_eq!(json, "\"base\"");

        let category = Category::AltEvm;
        let json = serde_json::to_string(&category).unwrap();
        assert_eq!(json, "\"alt_evm\"");
    }

    #[test]
    fn test_key_type_serialization() {
        let key_type = KeyType::Secp256k1;
        let json = serde_json::to_string(&key_type).unwrap();
        assert_eq!(json, "\"secp256k1\"");

        let key_type = KeyType::Ed25519;
        let json = serde_json::to_string(&key_type).unwrap();
        assert_eq!(json, "\"ed25519\"");
    }

    #[test]
    fn test_address_list_response() {
        let addresses = vec![
            Address::new(
                "wallet-1".to_string(),
                1,
                "BTC".to_string(),
                "Bitcoin".to_string(),
                0,
                "m/44'/0'/0'/0/0".to_string(),
                "address1".to_string(),
                Category::Base,
                KeyType::Secp256k1,
            ),
            Address::new(
                "wallet-1".to_string(),
                2,
                "ETH".to_string(),
                "Ethereum".to_string(),
                60,
                "m/44'/60'/0'/0/0".to_string(),
                "address2".to_string(),
                Category::Base,
                KeyType::Secp256k1,
            ),
        ];

        let response = AddressListResponse::new("wallet-1".to_string(), addresses);

        assert_eq!(response.wallet_id, "wallet-1");
        assert_eq!(response.addresses.len(), 2);
        assert_eq!(response.total_count, 2);
    }
}
